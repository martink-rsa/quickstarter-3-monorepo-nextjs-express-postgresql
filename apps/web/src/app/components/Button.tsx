'use client';

import { ButtonHTMLAttributes, ReactNode } from 'react';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children: ReactNode;
  variant?: 'primary' | 'secondary' | 'danger';
  size?: 'small' | 'medium' | 'large';
}

export function Button({
  children,
  variant = 'primary',
  size = 'medium',
  className = '',
  disabled = false,
  ...props
}: ButtonProps) {
  const baseClasses =
    'font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2';

  const variantClasses = {
    primary: 'bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 disabled:bg-blue-300',
    secondary:
      'bg-gray-200 text-gray-900 hover:bg-gray-300 focus:ring-gray-500 disabled:bg-gray-100',
    danger: 'bg-red-600 text-white hover:bg-red-700 focus:ring-red-500 disabled:bg-red-300',
  };

  const sizeClasses = {
    small: 'px-3 py-1.5 text-sm',
    medium: 'px-4 py-2 text-base',
    large: 'px-6 py-3 text-lg',
  };

  const combinedClasses = `${baseClasses} ${variantClasses[variant]} ${sizeClasses[size]} ${disabled ? 'cursor-not-allowed opacity-50' : 'cursor-pointer'} ${className}`;

  return (
    <button className={combinedClasses} disabled={disabled} {...props}>
      {children}
    </button>
  );
}
