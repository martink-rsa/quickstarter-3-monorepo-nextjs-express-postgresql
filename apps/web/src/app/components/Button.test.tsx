import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { Button } from './Button';

describe('Button', () => {
  it('renders children correctly', () => {
    render(<Button>Click me</Button>);
    expect(screen.getByRole('button')).toHaveTextContent('Click me');
  });

  it('applies primary variant classes by default', () => {
    render(<Button>Primary Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('bg-blue-600');
    expect(button).toHaveClass('text-white');
  });

  it('applies secondary variant classes when specified', () => {
    render(<Button variant="secondary">Secondary Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('bg-gray-200');
    expect(button).toHaveClass('text-gray-900');
  });

  it('applies danger variant classes when specified', () => {
    render(<Button variant="danger">Danger Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('bg-red-600');
    expect(button).toHaveClass('text-white');
  });

  it('applies small size classes when specified', () => {
    render(<Button size="small">Small Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('px-3');
    expect(button).toHaveClass('py-1.5');
    expect(button).toHaveClass('text-sm');
  });

  it('applies medium size classes by default', () => {
    render(<Button>Medium Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('px-4');
    expect(button).toHaveClass('py-2');
    expect(button).toHaveClass('text-base');
  });

  it('applies large size classes when specified', () => {
    render(<Button size="large">Large Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('px-6');
    expect(button).toHaveClass('py-3');
    expect(button).toHaveClass('text-lg');
  });

  it('handles click events', () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Click me</Button>);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('disables button when disabled prop is true', () => {
    render(<Button disabled>Disabled Button</Button>);
    const button = screen.getByRole('button');

    expect(button).toBeDisabled();
    expect(button).toHaveClass('cursor-not-allowed');
    expect(button).toHaveClass('opacity-50');
  });

  it('does not trigger click when disabled', () => {
    const handleClick = vi.fn();
    render(
      <Button disabled onClick={handleClick}>
        Disabled Button
      </Button>
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(handleClick).not.toHaveBeenCalled();
  });

  it('applies custom className', () => {
    render(<Button className="custom-class">Custom Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('custom-class');
  });

  it('passes through additional HTML button props', () => {
    render(
      <Button type="submit" aria-label="Submit form" data-testid="submit-button">
        Submit
      </Button>
    );

    const button = screen.getByRole('button');
    expect(button).toHaveAttribute('type', 'submit');
    expect(button).toHaveAttribute('aria-label', 'Submit form');
    expect(button).toHaveAttribute('data-testid', 'submit-button');
  });
});

describe('Button Snapshot Tests', () => {
  it('matches snapshot for primary button', () => {
    const { container } = render(<Button>Primary Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for secondary button', () => {
    const { container } = render(<Button variant="secondary">Secondary Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for danger button', () => {
    const { container } = render(<Button variant="danger">Danger Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for small button', () => {
    const { container } = render(<Button size="small">Small Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for large button', () => {
    const { container } = render(<Button size="large">Large Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for disabled button', () => {
    const { container } = render(<Button disabled>Disabled Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for button with all props', () => {
    const { container } = render(
      <Button variant="secondary" size="large" disabled className="custom-class">
        Complex Button
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
