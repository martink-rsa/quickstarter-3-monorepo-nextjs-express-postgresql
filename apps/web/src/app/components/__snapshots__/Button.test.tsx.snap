// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Snapshot Tests > matches snapshot for button with all props 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-gray-200 text-gray-900 hover:bg-gray-300 focus:ring-gray-500 disabled:bg-gray-100 px-6 py-3 text-lg cursor-not-allowed opacity-50 custom-class"
  disabled=""
>
  Complex Button
</button>
`;

exports[`Button Snapshot Tests > matches snapshot for danger button 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-red-600 text-white hover:bg-red-700 focus:ring-red-500 disabled:bg-red-300 px-4 py-2 text-base cursor-pointer "
>
  Danger Button
</button>
`;

exports[`Button Snapshot Tests > matches snapshot for disabled button 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 disabled:bg-blue-300 px-4 py-2 text-base cursor-not-allowed opacity-50 "
  disabled=""
>
  Disabled Button
</button>
`;

exports[`Button Snapshot Tests > matches snapshot for large button 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 disabled:bg-blue-300 px-6 py-3 text-lg cursor-pointer "
>
  Large Button
</button>
`;

exports[`Button Snapshot Tests > matches snapshot for primary button 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 disabled:bg-blue-300 px-4 py-2 text-base cursor-pointer "
>
  Primary Button
</button>
`;

exports[`Button Snapshot Tests > matches snapshot for secondary button 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-gray-200 text-gray-900 hover:bg-gray-300 focus:ring-gray-500 disabled:bg-gray-100 px-4 py-2 text-base cursor-pointer "
>
  Secondary Button
</button>
`;

exports[`Button Snapshot Tests > matches snapshot for small button 1`] = `
<button
  class="font-semibold rounded-lg transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 disabled:bg-blue-300 px-3 py-1.5 text-sm cursor-pointer "
>
  Small Button
</button>
`;
